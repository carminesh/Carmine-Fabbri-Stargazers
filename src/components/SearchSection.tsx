import { StyleSheet, Text, View } from 'react-native';
import React, { useMemo, useState } from 'react';
import ColorsPalette from '../constants/ColorsPalette';
import CustomTextInput from '../atoms/TextInput';
import SubmitButton from '../atoms/SubmitButton';

const SearchSection: React.FC = () => {
    const [user, setUser] = useState<string>('');
    const [repo, setRepo] = useState<string>('');

    /* memoized value to manage the button 'disabled' property */
    const isButtonEnabled: boolean = useMemo(() => {
        return user.length > 0 && repo.length > 0 ? false : true;
    }, [user, repo]);

    return (
        <View style={styles.viewContianer}>
            <CustomTextInput placeholder={'Username'} textValue={user} onChangeText={setUser} />
            <CustomTextInput placeholder={'Repository'} textValue={repo} onChangeText={setRepo} />
            <SubmitButton isButtonEnabled={isButtonEnabled} />
        </View>
    );
};

export default SearchSection;

const styles = StyleSheet.create({
    viewContianer: {
        paddingVertical: 16,
        width: '100%',
        height: 190,
        backgroundColor: ColorsPalette.LIGHT_BACKGROUND,
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'space-between',
    },
});
