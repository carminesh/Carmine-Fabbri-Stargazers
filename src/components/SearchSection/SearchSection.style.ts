import { StyleSheet } from 'react-native';
import ColorsPalette from '../../constants/ColorsPalette';

export const styles = StyleSheet.create({
    viewContianer: {
        paddingVertical: 16,
        width: '100%',
        height: 190,
        backgroundColor: ColorsPalette.LIGHT_BACKGROUND,
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'space-between',
    },
});
