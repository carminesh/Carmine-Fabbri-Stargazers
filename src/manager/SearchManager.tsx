import { Octokit } from '@octokit/core';
import Config from 'react-native-config';

const octokit = new Octokit({ auth: Config.GIT_HUB_ACCESS_TOKEN });

/* As you requested I followed API specs adherence from the official docs by using octokit. */
const getStargazersList = async (repoOwner: string, repoName: string) => {
    try {
        let response = await octokit.request(`GET /repos/${repoOwner}/${repoName}/stargazers?per_page=30`, {
            owner: repoOwner,
            repo: repoName,
        });

        return response.data;
    } catch (error: any) {
        console.error('Error: ', error.response.data.message);
    }
};

export { getStargazersList };
