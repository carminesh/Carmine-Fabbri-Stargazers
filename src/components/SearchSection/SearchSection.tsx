import { View } from 'react-native';
import React, { useCallback, useMemo, useState } from 'react';
import CustomTextInput from '../../atoms/TextInput/TextInput';
import SubmitButton from '../../atoms/SubmitButton/SubmitButton';
import { Stargazer } from '../../models/Stargazer';
import { getStargazersList } from '../../manager/SearchManager';
import { SEARCH_ICON, USER_ICON } from '../../assets/icons';
import { REPO_ICON } from '../../assets/icons/index';
import { useDispatch } from 'react-redux';
import { setStargazers } from '../../store/slices/StargazersSlice';
import { styles } from './SearchSection.list';

interface SearchSectionProps {
    setShowEplainer: React.Dispatch<React.SetStateAction<boolean>>;
    setIsLoading: React.Dispatch<React.SetStateAction<boolean>>;
}

const SearchSection: React.FC<SearchSectionProps> = ({ setShowEplainer, setIsLoading }) => {
    const [user, setUser] = useState<string>('nandorojo');
    const [repo, setRepo] = useState<string>('moti');
    const dispatch = useDispatch();

    /* memoized value to manage the button 'disabled' property */
    const isButtonEnabled: boolean = useMemo(() => {
        return user.length > 0 && repo.length > 0 ? false : true;
    }, [user, repo]);

    const fetchStargazers = useCallback(async () => {
        setIsLoading(true);
        setShowEplainer(false);
        try {
            let response = await getStargazersList(user, repo);

            /* In this section, we correctly map the endpoint response */
            if (response) {
                const mappedData: Stargazer[] = response.map((item: any) => ({
                    login: item.login,
                    id: item.id,
                    avatar_url: item.avatar_url,
                }));
                dispatch(setStargazers(mappedData));
            } else {
                /* here we set undefined when the user or the repo does not exist */
                dispatch(setStargazers(response));
            }
        } catch (e: unknown) {
            console.error('Error: ', e);
        } finally {
            setIsLoading(false);
        }
    }, [user, repo]);

    return (
        <View style={styles.viewContianer}>
            <CustomTextInput icon={USER_ICON} placeholder={'Search username'} textValue={user} onChangeText={setUser} />
            <CustomTextInput icon={REPO_ICON} placeholder={'Search repository'} textValue={repo} onChangeText={setRepo} />
            <SubmitButton icon={SEARCH_ICON} isButtonEnabled={isButtonEnabled} onPressCallback={fetchStargazers} />
        </View>
    );
};

export default SearchSection;

