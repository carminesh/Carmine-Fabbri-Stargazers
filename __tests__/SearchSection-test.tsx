import 'react-native';
import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import '@testing-library/jest-native/extend-expect';
import SearchSection from '../src/components/SearchSection/SearchSection';
import { getStargazersList } from '../src/manager/SearchManager';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';

jest.mock('../src/manager/SearchManager', () => ({
    getStargazersList: jest.fn(() =>
        Promise.resolve([
            { login: 'user1', id: 1, avatar_url: 'https://example.com/user1.jpg' },
            { login: 'user2', id: 2, avatar_url: 'https://example.com/user2.jpg' },
        ])
    ),
}));

describe('SearchSection', () => {
    const initialState = { output: 10 };
    const mockStore = configureStore();
    let store;

    it('renders correctly', () => {
        store = mockStore(initialState);
        const setShowEplainer = jest.fn();
        const setIsLoading = jest.fn();
        const tree = renderer
            .create(
                <Provider store={store}>
                    <SearchSection setShowEplainer={setShowEplainer} setIsLoading={setIsLoading} />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render the SearchSection', async () => {
        const setShowEplainer = jest.fn();
        const setIsLoading = jest.fn();
        store = mockStore(initialState);

        const { getByPlaceholderText, getByTestId } = render(
            <Provider store={store}>
                <SearchSection setShowEplainer={setShowEplainer} setIsLoading={setIsLoading} />
            </Provider>
        );

        const userInput = getByPlaceholderText('Search username');
        const repoInput = getByPlaceholderText('Search repository');
        const submitButton = getByTestId('submit-button');

        expect(userInput).toBeTruthy();
        expect(repoInput).toBeTruthy();
        expect(submitButton).toBeTruthy();
    });

    it('should change the input value and enable the submit button', async () => {
        const setShowEplainer = jest.fn();
        const setIsLoading = jest.fn();
        store = mockStore(initialState);

        const { getByPlaceholderText, getByTestId } = render(
            <Provider store={store}>
                <SearchSection setShowEplainer={setShowEplainer} setIsLoading={setIsLoading} />
            </Provider>
        );

        const userInput = getByPlaceholderText('Search username');
        const repoInput = getByPlaceholderText('Search repository');
        const submitButton = getByTestId('submit-button');

        fireEvent.changeText(userInput, 'test-user');
        fireEvent.changeText(repoInput, 'test-repo');

        expect(userInput.props.value).toBe('test-user');
        expect(repoInput.props.value).toBe('test-repo');
        expect(submitButton).not.toBeDisabled();
    });

    it('should call the getStargazersList function when submit button is pressed', async () => {
        store = mockStore(initialState);
        const setShowEplainer = jest.fn();
        const setIsLoading = jest.fn();
        const { getByTestId } = render(
            <Provider store={store}>
                <SearchSection setShowEplainer={setShowEplainer} setIsLoading={setIsLoading} />{' '}
            </Provider>
        );

        const userInput = getByTestId('user-input');
        fireEvent.changeText(userInput, 'nandorojo');
        const repoInput = getByTestId('repo-input');
        fireEvent.changeText(repoInput, 'moti');
        const submitButton = getByTestId('submit-button');
        fireEvent.press(submitButton);
        expect(getStargazersList).toHaveBeenCalledWith('nandorojo', 'moti');
        expect(setIsLoading).toHaveBeenCalledWith(true);
        expect(setShowEplainer).toHaveBeenCalledWith(false);
    });
});
