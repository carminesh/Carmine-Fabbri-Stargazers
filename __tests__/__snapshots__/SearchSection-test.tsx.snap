// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSection renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "column",
      "height": 190,
      "justifyContent": "space-between",
      "paddingVertical": 16,
      "width": "100%",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#a6a6a6",
          "borderRadius": 5,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 45,
          "justifyContent": "center",
          "paddingHorizontal": 20,
          "width": 718,
        },
        false,
      ]
    }
    testID="text-input-container"
  >
    <Image
      resizeMode="contain"
      source={
        {
          "testUri": "../../../src/assets/icons/userIcon.png",
        }
      }
      style={
        [
          {
            "height": 15,
            "marginRight": 10,
            "tintColor": "#a6a6a6",
            "width": 15,
          },
          false,
        ]
      }
      testID="custom-text-input-icon"
    />
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Search username"
      style={
        {
          "height": "70%",
          "width": "100%",
        }
      }
      testID="user-input"
      value="nandorojo"
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#a6a6a6",
          "borderRadius": 5,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 45,
          "justifyContent": "center",
          "paddingHorizontal": 20,
          "width": 718,
        },
        false,
      ]
    }
    testID="text-input-container"
  >
    <Image
      resizeMode="contain"
      source={
        {
          "testUri": "../../../src/assets/icons/repoIcon.png",
        }
      }
      style={
        [
          {
            "height": 15,
            "marginRight": 10,
            "tintColor": "#a6a6a6",
            "width": 15,
          },
          false,
        ]
      }
      testID="custom-text-input-icon"
    />
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Search repository"
      style={
        {
          "height": "70%",
          "width": "100%",
        }
      }
      testID="repo-input"
      value="moti"
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2C68F7",
        "borderRadius": 5,
        "flexDirection": "row",
        "height": 45,
        "justifyContent": "center",
        "opacity": 1,
        "width": 718,
      }
    }
    testID="submit-button"
  >
    <Image
      resizeMode="contain"
      source={
        {
          "testUri": "../../../src/assets/icons/searchIcon.png",
        }
      }
      style={
        {
          "height": 15,
          "marginRight": 10,
          "tintColor": "#FFFFFF",
          "width": 15,
        }
      }
      testID="submit-button-icon"
    />
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontWeight": "600",
        }
      }
    >
      Search
    </Text>
  </View>
</View>
`;
